design-proposals/klip-47-non-null-columns-and-component-types.md

```diff
@@ -0,0 +1,183 @@
+```
+
+Given data in the pageviews topic:
+```sql
+3,,
+3,user,page
+2,,page1
+null,user,page
+```
+
+A `SELECT` query from the table would return only the two records that didn't violate the `NOT NULL` constraint:
+```sql
+ksql> select * from pageviews emit changes;
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|VIEWTIME       |USERID                                                          |PAGEID                                                          |
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|3              |user                                                            |page                                                            |
+|null           |user                                                            |page1                                                           |
+```
+
+Declaring a column as `NOT NULL` in a CSAS/CTAS would look like:
+```sql
+CREATE stream pageviews_no_null as select CAST(viewtime AS BIGINT NOT NULL), userid, pageid from pageviews;
+```
+The resulting stream's columns would all be `NOT NULL`
+
+```sql
+ksql> select * from pageviews_no_null emit changes;
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|VIEWTIME       |USERID                                                          |PAGEID                                                          |
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|3              |user                                                            |page                                                            |
+```
+
+Attempting to insert invalid records would error out:
+```sql
+ksql> INSERT INTO pageviews VALUES (5, NULL, 'page23');
+Failed to insert values into 'PAGEVIEWS': null violates the NOT NULL constraint for field "USERID"
+
+ksql> INSERT INTO pageviews VALUES (NULL, 'user3', 'page23'); # this insert succeeds
+
+ksql> INSERT INTO pageviews_no_null VALUES (NULL, 'user3', 'page23');
+Failed to insert values into 'PAGEVIEWS': null violates the NOT NULL constraint for field "VIEWTIME"
+```
+
+`NOT NULL` for arrays, maps, struct
+```sql
+CREATE STREAM valid_data (arr ARRAY<BIGINT NOT NULL>, mapping MAP<VARCHAR NOT NULL, BIGINT NOT NULL>, valid_struct STRUCT<c DOUBLE NOT NULL, d BOOLEAN>) ...
+```
+Records that have column values such as
+* `{arr: [3,4,null] ... }`
+* `{mapping: {"key": null} ...}`
+* `{mapping: {null: 5} ...}`
+* `{valid_struct: {c: null, d: true} ...}`
+
+Would be dropped when reading from the `valid_data` stream and trying to insert record with these column values would error out.
```

Comment:
Maybe we can add another example here illustrating how `NOT NULL` would effect for sink topics. E.g. my understanding is that if we have a `CTAS` from aggregating a stream with the aggregating column defined as `NOT NULL`, and user's UDAF happens to set the return value as null, then when reading from this table we would not see that result from UDAF as well.

Reply:
I added another example with an aggregate function